statistics/statistics.go

```diff
@@ -21,329 +21,143 @@ package statistics
-		case "98%_ile":
-			val, err = stats.PercentileNearestRank(buffer, 98)
-		case "minimum":
-			val, err = min, minErr
-		case "maximum":
-			val, err = max, maxErr
-		case "range":
-			val = max - min
-		case "mode":
-			modeVal, err = stats.Mode(buffer)
-		case "kurtosis":
-			val, err = p.Kurtosis(buffer)
-		case "skewness":
-			val, err = p.Skewness(buffer)
-		case "sum":
-			val, err = stats.Sum(buffer)
-		case "trimean":
-			val, err = stats.Trimean(buffer)
-		case "quartile_range":
-			val, err = stats.InterQuartileRange(buffer)
-		default:
-			st := fmt.Sprintf("Unknown statistic received %T:", stat)
-			log.Errorf(st)
-			err = errors.New(st)
+// Process processes the data, inputs the data into sorted buffer and calls the GetStats method
+func (p *Plugin) Process(metrics []plugin.Metric, cfg plugin.Config) ([]plugin.Metric, error) {
+	var result []plugin.Metric
+	for _, metric := range metrics {
+		slidingWindowLength, slidingFactor, stats, err := GetConfig(metric.Config)
+		if err != nil {
+			return nil, err
 		}
 
+		// convert any number to float64
+		floatValue, err := dataToFloat64(metric.Data)
 		if err != nil {
-			log.Warnf("Error in %T", stat)
+			return nil, err
 		}
 
-		metric := plugin.Metric{
-			Data:      val,
-			Namespace: plugin.NewNamespace(namespace, stat),
-			Timestamp: time,
-			Unit:      unit,
-			Tags:      tags,
-		}
+		nsSlice := metric.Namespace.Strings()
+		ns := strings.Join(nsSlice, "")
+		_, ok := p.buffer[ns]
+		if !ok {
+			//buffer doesn't exist, so create one
+			p.buffer[ns] = &dataBuffer{
+				data:            make([]*data, 0, slidingWindowLength),
+				dataByTimestamp: make([]*data, 0, slidingWindowLength)}
 
-		if stat == "mode" {
-			metric.Data = modeVal
+		} else {
+			if slidingWindowLength != cap(p.buffer[ns].data) {
+				// TODO: test if buffer size from the config is different than cap(p.buffer[ns])
+			}
 		}
 
-		result = append(result, metric)
-
+		p.buffer[ns].Insert(floatValue, metric.Timestamp)
+		// add a new element to the sorted list
+		if p.buffer[ns].slidingFactorIndex%slidingFactor == 0 {
+			mts, err := p.buffer[ns].GetStats(stats, nsSlice)
+			if err != nil {
+				return nil, err
+			}
+			result = append(result, mts...)
+		}
+		p.buffer[ns].slidingFactorIndex++
 	}
-	return result, err
+	return result, nil
 }
 
-//Calaculates the mean and standard deviation of a float64 array.
-func (p *Plugin) MeanStdDev(buffer []float64) (float64, float64, error) {
-	mean, err := stats.Mean(buffer)
+// GetConfig returns the config policy
+func GetConfig(cfg plugin.Config) (slidingWinLen, slidingFac int, statistics []string, err error) {
+	var stats string
+	stats, err = cfg.GetString("statistics")
 	if err != nil {
-		log.Warn(err)
-		return math.NaN(), math.NaN(), err
+		err = fmt.Errorf("%v: \"statistics\"", err)
+		return
 	}
+	statistics = strings.Split(stats, ",")
+	var tmp int64
 
-	stdev, err := stats.StandardDeviation(buffer)
+	tmp, err = cfg.GetInt("slidingWindowLength")
```

Comment:
No, it's just a comment, It's fine in different style as we don't have standard now.